File: backend/app/schemas/investment_evidence.py
"""Source facts and reviewed links; decimal strings are part of the API contract."""
from datetime import date, datetime
from decimal import Decimal, DecimalException
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class EvidenceLegInput(BaseModel):
    model_config = ConfigDict(extra="forbid")

    key: str = Field(min_length=1, max_length=255)
    asset_symbol: str | None = Field(None, max_length=64)
    asset_id: UUID | None = None
    chain: str | None = Field(None, max_length=64)
    token_address: str | None = Field(None, max_length=255)
    isin: str | None = Field(None, max_length=32)
    provider_asset_id: str | None = Field(None, max_length=255)
    direction: Literal["in", "out", "unknown"] = "unknown"
    classification: str = Field("unknown", max_length=64)
    quantity: Decimal | None = None
    unit_price: Decimal | None = None
    execution_currency: str | None = Field(None, max_length=64)
    unit_price_origin: Literal["reported", "derived_execution", "derived_spot", "unknown"] = "unknown"
    subtotal: Decimal | None = None
    total: Decimal | None = None
    fee: Decimal | None = None
    fee_currency: str | None = Field(None, max_length=64)
    valuation_currency: str | None = Field(None, max_length=64)
    valuation_amount: Decimal | None = None
    external_funding_amount: Decimal | None = None
    external_funding_currency: str | None = Field(None, max_length=64)
    acquisition_basis: Decimal | None = None
    transaction_ref: str | None = Field(None, max_length=255)
    leg_ref: str | None = Field(None, max_length=255)
    execution_id: str | None = Field(None, max_length=255)

    @field_validator(
        "quantity", "unit_price", "subtotal", "total", "fee",
        "external_funding_amount", "acquisition_basis", "valuation_amount", mode="before",
    )
    @classmethod
    def exact_decimal(cls, value):
        if isinstance(value, float):
            raise ValueError("Financial amounts must be decimal strings, not JSON floats")
        if value is not None:
            try:
                number = Decimal(value)
            except (DecimalException, TypeError, ValueError):
                raise ValueError("Financial amounts must be valid decimal strings") from None
            if not number.is_finite():
                raise ValueError("Financial amounts must be finite")
            if number < 0:
                raise ValueError("Amounts are unsigned; direction records the sign")
            if len(number.as_tuple().digits) > 128 or (number and abs(number.adjusted()) > 128):
                raise ValueError("Source amount exceeds the supported 128-digit arithmetic range")
        return value


class EvidenceObservationInput(BaseModel):
    model_config = ConfigDict(extra="forbid")

    reference: str = Field(min_length=1, max_length=512)
    source_reference: str | None = Field(None, max_length=512)
    source: str = Field(min_length=1, max_length=64)
    source_kind: Literal[
        "primary_activity", "balance_snapshot", "remaining_lots", "tax_workpaper", "recovery_notice"
    ] = "primary_activity"
    provider: str = Field(min_length=1, max_length=64)
    source_account_id: str | None = Field(None, max_length=255)
    account_external_id: str | None = Field(None, max_length=255)
    holding_external_id: str | None = Field(None, max_length=255)
    source_local_id: str | None = Field(None, max_length=255)
    source_locator: str = Field(min_length=1, max_length=512)
    observed_at: datetime | None = None
    event_time_raw: str | None = Field(None, max_length=255)
    event_date: date | None = None
    event_at: datetime | None = None
    timezone: str | None = Field(None, max_length=64)
    time_precision: Literal["unknown", "date", "minute", "second", "fractional"] = "unknown"
    provider_status: str | None = Field(None, max_length=64)
    network_status: str | None = Field(None, max_length=64)
    settlement_status: Literal["settled", "pending", "failed", "unknown"] = "unknown"
    order_ref: str | None = Field(None, max_length=255)
    historical_workspace_label: str | None = Field(None, max_length=255)
    coverage: list[str] = Field(default_factory=list, max_length=100)
    reason_codes: list[str] = Field(default_factory=list, max_length=100)
    source_fields: dict[str, str | None] = Field(default_factory=dict)
    legs: list[EvidenceLegInput] = Field(min_length=1, max_length=100)

    @field_validator("event_at", "observed_at")
    @classmethod
    def aware_time(cls, value):
        if value is not None and value.utcoffset() is None:
            raise ValueError("An instant requires a timezone; retain unresolved time as event_time_raw")
        return value

    @model_validator(mode="after")
    def unique_legs(self):
        if len({leg.key for leg in self.legs}) != len(self.legs):
            raise ValueError("Source leg keys must be distinct within an observation")
        if self.time_precision in {"date", "unknown"} and self.event_at is not None:
            raise ValueError("Date-only and unknown observations cannot claim an instant")
        return self

    @field_validator("source_fields")
    @classmethod
    def bounded_source_fields(cls, value):
        allowed = {
            "acquisition_date", "disposal_date", "acquisition_basis", "proceeds",
            "quantity", "unit_price", "subtotal", "total", "fee", "fee_currency",
            "valuation_currency", "valuation_amount", "external_funding_amount",
            "external_funding_currency", "classification",
            "execution_currency", "unit_price_currency", "subtotal_currency", "total_currency", "unit_price_origin",
        }
        if value.keys() - allowed or any(v is not None and len(v) > 512 for v in value.values()):
            raise ValueError("Only bounded, named source evidence fields may be retained")
        return value


class EvidenceAllocation(BaseModel):
    leg_id: UUID
    quantity: Decimal | None = None

    @field_validator("quantity", mode="before")
    @classmethod
    def exact_decimal(cls, value):
        return EvidenceLegInput.exact_decimal(value)


class EvidenceDecision(BaseModel):
    model_config = ConfigDict(extra="forbid")

    observation_ref: str
    leg_key: str
    action: Literal["retain", "apply", "link"] = "retain"
    allocations: list[EvidenceAllocation] = Field(default_factory=list, max_length=100)
    reason: str | None = Field(None, max_length=500)
    settlement_confirmed: bool = False


class EvidenceOpeningBoundary(BaseModel):
    as_of: date
    overlap_reviewed: bool
    assumption: str = Field(min_length=1, max_length=500)


class EvidenceTarget(BaseModel):
    workspace_id: UUID
    workspace_name: str
    group_id: UUID
    group_name: str
    account_id: UUID | None = None


class EvidenceSourceRef(BaseModel):
    observation_ref: str
    source: str
    source_local_id: str | None = None
    source_locator: str
    leg_key: str


class EvidenceCandidate(BaseModel):
    leg_id: UUID
    event_id: UUID
    asset_symbol: str | None = None
    direction: str
    classification: str
    quantity: Decimal | None = None
    event_date: date | None = None
    source_refs: list[EvidenceSourceRef] = Field(default_factory=list)


class EvidenceEffects(BaseModel):
    ledger_rows: int = 0
    units_delta: Decimal | None = Decimal("0")
    basis_delta: Decimal | None = None


class EvidenceLinkedLeg(BaseModel):
    link_id: UUID
    leg: EvidenceCandidate
    quantity: Decimal | None = None


class EvidenceRecord(BaseModel):
    observation_ref: str
    leg_key: str
    match_status: Literal["linked", "candidate", "conflicting", "unmatched"]
    application_status: Literal["already_applied", "eligible", "blocked", "not_applicable"]
    source_refs: list[EvidenceSourceRef] = Field(default_factory=list)
    candidate_legs: list[EvidenceCandidate] = Field(default_factory=list)
    link_ids: list[UUID] = Field(default_factory=list)
    links: list[EvidenceLinkedLeg] = Field(default_factory=list)
    reason_codes: list[str] = Field(default_factory=list)
    conflicting_fields: list[str] = Field(default_factory=list)
    effects: EvidenceEffects = Field(default_factory=EvidenceEffects)


class EvidenceReconciliation(BaseModel):
    asset_symbol: str | None = None
    chain: str | None = None
    token_address: str | None = None
    provider_asset_id: str | None = None
    isin: str | None = None
    opening_quantity: Decimal | None = None
    opening_assumption: str = "unknown"
    opening_as_of: date | None = None
    snapshot_quantity: Decimal | None = None
    snapshot_as_of: str | None = None
    settled_movement_quantity: Decimal = Decimal("0")
    expected_closing_quantity: Decimal | None = None
    discrepancy: Decimal | None = None
    missing_coverage: list[str] = Field(default_factory=list)
    unresolved_fee_semantics: bool = True
    unresolved_funding_semantics: bool = True
    basis_complete: bool = False
    history_complete: bool = False


class EvidencePreview(BaseModel):
    revision: str
    target: EvidenceTarget
    observations: list[EvidenceObservationInput]
    records: list[EvidenceRecord]
    reconciliation: list[EvidenceReconciliation] = Field(default_factory=list)


class EvidenceConfirmRequest(BaseModel):
    group_id: UUID
    decisions: list[EvidenceDecision]
    expected_revision: str
    opening_boundary: EvidenceOpeningBoundary | None = None
    allow_unpriced: bool = False


class EvidenceResult(BaseModel):
    import_log_id: UUID | None = None
    imported: int = 0
    retained: int = 0
    linked: int = 0
    evidence: EvidencePreview

File: frontend/src/types/investment-evidence.ts
/** Financial values remain decimal strings from the source to the server. */
export interface EvidenceLeg {
  key: string
  asset_symbol: string | null
  asset_id: string | null
  chain: string | null
  token_address: string | null
  isin: string | null
  provider_asset_id?: string | null
  direction: 'in' | 'out' | 'unknown'
  classification: string
  quantity: string | null
  unit_price: string | null
  execution_currency?: string | null
  unit_price_origin?: string | null
  subtotal: string | null
  total: string | null
  fee: string | null
  fee_currency: string | null
  valuation_currency: string | null
  valuation_amount: string | null
  external_funding_amount: string | null
  external_funding_currency: string | null
  acquisition_basis: string | null
  transaction_ref: string | null
  leg_ref: string | null
  execution_id: string | null
  token_program?: string | null
  source_address?: string | null
  destination_address?: string | null
  source_owner?: string | null
  destination_owner?: string | null
  raw_units?: string | null
  decimals?: number | null
  quantity_role?: string | null
  fee_payer?: string | null
  fee_semantics?: 'separate' | 'none' | 'included' | 'unknown'
  derivation?: Record<string, string>
}

export type EvidenceSourceKind = 'primary_activity' | 'balance_snapshot' | 'remaining_lots' | 'tax_workpaper' | 'recovery_notice'

export interface EvidenceObservation {
  reference: string
  source_reference?: string | null
  source: string
  source_kind: EvidenceSourceKind
  provider: string
  source_account_id: string | null
  account_external_id?: string | null
  holding_external_id?: string | null
  source_local_id: string | null
  source_locator: string
  observed_at: string | null
  event_time_raw: string | null
  event_date: string | null
  event_at: string | null
  timezone: string | null
  time_precision: 'unknown' | 'date' | 'minute' | 'second' | 'fractional'
  provider_status: string | null
  network_status: string | null
  settlement_status: 'settled' | 'pending' | 'failed' | 'unknown'
  order_ref: string | null
  historical_workspace_label: string | null
  coverage: string[]
  reason_codes?: string[]
  source_fields?: Record<string, string | null>
  legs: EvidenceLeg[]
}

export interface EvidenceSourceRef {
  observation_ref: string
  source: string
  source_local_id: string | null
  source_locator: string
  leg_key: string
}

export interface EvidenceCandidate {
  leg_id: string
  event_id: string
  asset_symbol: string | null
  direction: string
  classification: string
  quantity: string | null
  event_date: string | null
  source_refs: EvidenceSourceRef[]
}

export interface EvidenceRecord {
  observation_ref: string
  leg_key: string
  match_status: 'linked' | 'candidate' | 'conflicting' | 'unmatched'
  application_status: 'already_applied' | 'eligible' | 'blocked' | 'not_applicable'
  source_refs: EvidenceSourceRef[]
  candidate_legs: EvidenceCandidate[]
  link_ids: string[]
  links?: { link_id: string; leg: EvidenceCandidate; quantity: string | null }[]
  reason_codes: string[]
  conflicting_fields: string[]
  effects: { ledger_rows: number; units_delta: string | null; basis_delta: string | null }
}

export interface EvidencePreview {
  revision: string
  target: { workspace_id: string; workspace_name: string; group_id: string; group_name: string; account_id: string | null }
  observations: EvidenceObservation[]
  records: EvidenceRecord[]
  reconciliation: {
    asset_symbol: string | null
    chain?: string | null
    token_address?: string | null
    provider_asset_id?: string | null
    isin?: string | null
    opening_quantity: string | null
    opening_assumption: string
    opening_as_of: string | null
    snapshot_quantity: string | null
    snapshot_as_of: string | null
    settled_movement_quantity: string
    expected_closing_quantity?: string | null
    discrepancy?: string | null
    missing_coverage: string[]
    unresolved_fee_semantics: boolean
    unresolved_funding_semantics: boolean
    basis_complete: boolean
    history_complete: boolean
  }[]
}

export interface EvidenceDecision {
  observation_ref: string
  leg_key: string
  action: 'retain' | 'apply' | 'link'
  allocations?: { leg_id: string; quantity?: string | null }[]
  reason?: string
  settlement_confirmed?: boolean
}

export interface EvidenceOpeningBoundary {
  as_of: string
  overlap_reviewed: boolean
  assumption: string
}

export interface EvidenceResult {
  import_log_id: string | null
  imported: number
  retained: number
  linked: number
  evidence: EvidencePreview
}

export interface EvidenceImportOptions {
  mode?: 'orders' | 'evidence' | 'opening_lots'
  column_mapping?: Record<string, string>
  date_format?: string
  group_id?: string | null
  allow_unpriced?: boolean
  provider?: string
  source_account_id?: string
  source_kind?: EvidenceSourceKind
  connection_id?: string | null
  opening_boundary?: EvidenceOpeningBoundary
}
